src/core/pinned_memory_manager.cc

```diff
@@ -50,60 +51,102 @@ PointerToString(void* ptr)
 
 std::unique_ptr<PinnedMemoryManager> PinnedMemoryManager::instance_;
 
+class PinnedMemoryManagerImpl : public PinnedMemoryManager {
+ public:
+  static Status Create(
+      std::unique_ptr<PinnedMemoryManager>* manager, const Options& options);
+
+  ~PinnedMemoryManagerImpl();
+
+  Status AllocInternal(
+      void** ptr, uint64_t size,
+      bool allow_nonpinned_fallback = false) override;
+  Status FreeInternal(void* ptr) override;
+
+ private:
+  PinnedMemoryManagerImpl(void* pinned_memory_buffer, uint64_t size);
+  std::mutex info_mtx_;
+  std::map<void*, std::pair<bool, uint64_t>> memory_info_;
+
+  void* pinned_memory_buffer_;
+  std::mutex buffer_mtx_;
+  boost::interprocess::managed_external_buffer managed_pinned_memory_;
+};
+
 Status
-PinnedMemoryManager::Create(const Options& options)
+PinnedMemoryManagerImpl::Create(
+    std::unique_ptr<PinnedMemoryManager>* manager, const Options& options)
 {
-  if (instance_ != nullptr) {
-    return Status(
-        RequestStatusCode::ALREADY_EXISTS,
-        "PinnedMemoryManager has been created");
+  void* buffer = nullptr;
+#ifdef TRTIS_ENABLE_GPU
+  auto err = cudaHostAlloc(
+      &buffer, options.pinned_memory_pool_byte_size_, cudaHostAllocPortable);
+  if (err != cudaSuccess) {
+    buffer = nullptr;
+    LOG_ERROR << "failed to allocate pinned system memory: "
+              << std::string(cudaGetErrorString(err));
+  } else {
+    LOG_VERBOSE(1) << "Pinned memory pool is created at '"
+                   << PointerToString(buffer) << "' with size "
+                   << options.pinned_memory_pool_byte_size_;
   }
-
-  instance_.reset(new PinnedMemoryManager(options));
+#endif  // TRTIS_ENABLE_GPU
+  manager->reset(new PinnedMemoryManagerImpl(
+      buffer, options.pinned_memory_pool_byte_size_));
   return Status::Success;
 }
 
-Status
-PinnedMemoryManager::Alloc(
-    void** ptr, uint64_t size, bool allow_nonpinned_fallback)
+PinnedMemoryManagerImpl::PinnedMemoryManagerImpl(
+    void* pinned_memory_buffer, uint64_t size)
+    : pinned_memory_buffer_(pinned_memory_buffer)
 {
-  if (instance_ == nullptr) {
-    return Status(
-        RequestStatusCode::UNAVAILABLE,
-        "PinnedMemoryManager has not been created");
+  if (pinned_memory_buffer_ != nullptr) {
+    managed_pinned_memory_ = boost::interprocess::managed_external_buffer(
+        boost::interprocess::create_only_t{}, pinned_memory_buffer_, size);
+    // Sanity check on whether we need to configure 'size_type'
```

Comment:
What check is this comment referring to?

Reply:
This check can be removed. This is for debugging purpose, Boost uses 'size_type' which is obscured to me, so print the size seen by Boost to make sure it is larger than 32 bits.